services/addons/images/firmware_manager/firmware_manager.py

```diff
@@ -271,6 +302,53 @@ def check_for_upgrades():
         start_tasks_from_queue()
 
 
+def was_latest_ping_successful_for_rsu(rsu_ip):
+    query = f"select result from ping where rsu_id=(select rsu_id from rsus where ipv4_address='{rsu_ip}') order by timestamp desc limit 1"
+    latest_ping_successful = pgquery.query_db(query)[0][0]
+    logging.info(f"Latest ping result for '{rsu_ip}': {latest_ping_successful}")
+    return latest_ping_successful
+
+
+def increment_consecutive_failure_count_for_rsu(rsu_ip):
+    if not does_consecutive_failure_count_exist_for_rsu(rsu_ip):
+        insertion_query = f"insert into consecutive_firmware_upgrade_failures (rsu_id, consecutive_failures) values ((select rsu_id from rsus where ipv4_address='{rsu_ip}'), 1)"
+        pgquery.write_db(insertion_query)
+    else:
+        increment_query = f"update consecutive_firmware_upgrade_failures set consecutive_failures=consecutive_failures+1 where rsu_id=(select rsu_id from rsus where ipv4_address='{rsu_ip}')"
+        pgquery.write_db(increment_query)

```

Comment:
question (non-blocking): I'm new-ish to this codebase and a little rusty with my Postgres, but I'm wondering if there's a reason we don't use an [Upsert ](https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-upsert/)? If we used an upsert statement we could combine the logic in `does_consecutive_failure_count_exist_for_rsu`, the if-block, and the else-block in this function into one call to the database.

Reply:
I really like the idea of using ON CONFLICT to reduce the number of database queries here. I've opened a PR to implement your suggestion here: https://github.com/CDOT-CV/jpo-cvmanager/pull/112